grub-core/loader/multiboot_elfxx.c

```diff
@@ -101,30 +106,96 @@ CONCAT(grub_multiboot_load_elf, XX) (mbi_load_data_t *mld)
     {
       load_size = highest_load - mld->link_base_addr;
 
-      grub_dprintf ("multiboot_loader", "align=0x%lx, preference=0x%x, "
+      grub_dprintf ("multiboot_loader", "link_base_addr=0x%08x, align=0x%lx, preference=0x%x, "
 		    "load_size=0x%x, avoid_efi_boot_services=%d\n",
-		    (long) mld->align, mld->preference, load_size,
+		    mld->link_base_addr, (long) mld->align, mld->preference, load_size,
 		    mld->avoid_efi_boot_services);
 
-      if (load_size > mld->max_addr || mld->min_addr > mld->max_addr - load_size)
+      if (grub_slaunch_platform_type () == SLP_INTEL_TXT)
+	{
+#ifndef GRUB_USE_MULTIBOOT2
+	  return grub_error (GRUB_ERR_BAD_OS, "Only multiboot2 supported for slaunch");
+#else
+	  /*
+	   * We allocate the the binary together with page tables to make one
+	   * contiguous block for MLE. We have to align up to PMR (2MB).
+	   */
+	  total_size = ALIGN_UP(load_size, GRUB_TXT_PMR_ALIGN);
+
+	  slparams->mle_size = total_size;
+
+	  slparams->mle_ptab_size = grub_txt_get_mle_ptab_size (total_size);
+	  slparams->mle_ptab_size = ALIGN_UP (slparams->mle_ptab_size, GRUB_TXT_PMR_ALIGN);
+
+	  /* Do not go below GRUB_TXT_PMR_ALIGN. */
+	  if (mld->align < GRUB_TXT_PMR_ALIGN)
+	    mld->align = GRUB_TXT_PMR_ALIGN;
+
+	  mld->min_addr = (mld->min_addr > slparams->mle_ptab_size) ?
+			  (mld->min_addr - slparams->mle_ptab_size) : mld->align;
```

Comment:
This line moves min_addr lower, why?

Reply:
To avoid situations where there is no space for page tables below the ELF image

Reply:
But page tables are allocated later, they are not included in this chunk as far as I can tell.

Reply:
Yes, but does that mean we cannot ensure there will be space for page tables here? What is the point if we will fail with page tables allocation later 100% of times for given binary?

Reply:
Then you should add to `min_addr`, not subtract from it.

Reply:
Removed it completely. Still boots